### PR TITLE
Make deriver

### PR DESCRIPTION
**Pull request by ayc9**
This is an implementation of the `make` PPX deriver, based on previous implementations and updated using latest `ppxlib` API. The deriver works on record types, and supports `@main` and `@default` annotations on specific fields within the record type. Usage is documented in the readme. Special attention is needed for:

- General code structure and readability (changes to function names, refactors welcome)
- Testing coverage (especially for cases with colluding features, like `@main` on an `option` type)

Thank you!

### CODE REVIEW

File: src/make/ppx_make.ml [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+      let main_labels, labels = List.fold_left (fun (main_labels, labels) label ->
+        match Attribute.get main_attr label with
+        | Some _ -> label::main_labels, labels
+        | None -> main_labels, label :: labels
+      ) ([], []) labels in 
+      match main_labels with 
+      | [] -> Ok (None, labels)
+      | [ main_label ] -> Ok (Some main_label, labels)
+      | main_labels -> Error (List.map(fun ({ pld_loc; _ }) -> 
+        Location.error_extensionf ~loc:pld_loc "Duplicate [@deriving.%s.main] annotation" "make"
```
Comment (pitag-ha) [inline src/make/ppx_make.ml]: Just out of curiosity: Why have you extracted `"make"` instead of

```suggestion
        Location.error_extensionf ~loc:pld_loc "Duplicate [@deriving.make.main] annotation" 
```

? Do you want to re-use this function in the context of another deriver as well?

File: src/make/ppx_make.ml [hunk 2]
```diff
@@ -0,0 +1,247 @@
+(* Generated code should depend on the environment in scope as little as
+   possible.  E.g. rather than [foo = []] do [match foo with [] ->], to 
+   eliminate the use of [=], which might be overwritten in the environment.
+   It is especially important to not use polymorphic comparisons. *)
```
Comment (pitag-ha) [inline src/make/ppx_make.ml]: This is definitely true and important! I'm just wondering if we should move that to the README into the section *hygienic base ecosystem* as an example of what we mean by "hygienic"? By the end of the day, this is true for every deriver, not just for `make`. What do you think?

File: src/make/ppx_make.ml [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+   open Ast_builder.Default
+
+   module Annotations = struct 
+    let default_attr = 
+      Attribute.declare 
+        "standard_derivers.make.default" 
+        Attribute.Context.label_declaration
+        Ast_pattern.(single_expr_payload __)
+        (fun expr -> expr)
+    ;;
```
Comment (pitag-ha) [inline src/make/ppx_make.ml]: If that's ok with you, I think it would be nice to remove these double semi-colons throughout the file. I know that can be considered just a matter of style, but I do have reasons to remove them:  in other programming languages semicolons or double-semicolons are actually needed to mark the end of an expression, but in OCaml they aren't. So often, when folks new to OCaml come across double semicolons in OCaml code, they get confused asking themselves "Do I need double semicolons? Don't I need them? When do I need them?" etc. So I think it's a good style to avoid them unless actually needed, which is very seldom (I think just in `utop`, or to separate different structure items that consist of a floating expression, which I've only seen used in very rare cases). Does it sound good to you to remove them?

File: src/make/ppx_make.ml [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+      ) main_labels )
+    ;;
+   end
+
+   module Check = struct
+    let is_derivable ~loc rec_flag tds =
+      match rec_flag with
+       | Nonrecursive ->
+         Error (Location.error_extensionf ~loc "nonrec is not compatible with the `make' preprocessor.")
+       | _ ->
```
Comment (pitag-ha) [inline src/make/ppx_make.ml:38]: Oh, that's very interesting! Was it a deliberate choice of yours to use the wildcard here instead of of `Recursive`?

The reason why I'm asking is that usually in OCaml, it's considered a good habit to write out the pattern and to avoid the wildcard whenever that's easy to do. That way you allow the compiler to check for exhaustiveness. So in most situations, I'd make the following suggestion:

```suggestion
       | Recursive ->
```

However, you're in a very peculiar situation of writing OCaml here: you're touching code without a stability guarantee since the AST tends to get modified a bit with every compiler release. Concretely, it's possible that at some point in the future, the type
```
type rec_flag = Nonrecursive | Recursive
```
will be modified to, let's say
```
type rec_flag = Nonrecursive | Recursive of int
```
or
```
type rec_flag = Nonrecursive | Recursive | Tailrecursive
```
(obviously, I'm totally making things up here. I'm just trying to give an idea of what kind of changes might be made to that type.)

And if that happens, the choice of using the wildcard here will turn out very beneficial since in that case your code won't break (whereas if you wrote out `Recursive` and let the compiler check for exhaustiveness, then the code would break)!

So, long story short: I like the fact that you're using the wildcard here. I just wanted to explain a bit _why_ I like it since it's not the "common good habit".

File: src/make/ppx_make.ml [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+   module Construct = struct
+     (* Additional AST construction helpers *)
+   
+     let apply_type ~loc ~ty_name ~tps = 
+      ptyp_constr ~loc (Located.lident ~loc ty_name) tps
+     ;;
+
+     let lambda ~loc patterns body =
+      List.fold_left (fun acc (lab, pat, default) ->
+       pexp_fun ~loc lab default pat acc) body patterns
```
Comment (pitag-ha) [inline src/make/ppx_make.ml]: I'm thinking if this might be a bit easier to read if you used the same order of arguments and the same nomenclature as in [parsetree.mli](https://github.com/ocaml/ocaml/blob/trunk/parsing/parsetree.mli#L284):
```suggestion
      List.fold_left (fun expr (arg_label, default, pat) ->
       pexp_fun ~loc arg_label default pat expr) body patterns
```

What do you think? Do you think that might help?

File: src/make/ppx_make.ml [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+     ;;
+  
+     let lambda_sig ~loc arg_tys body_ty =
+      List.fold_left (fun acc (lab, arg_ty) ->
+       ptyp_arrow ~loc lab arg_ty acc) body_ty arg_tys
```
Comment (pitag-ha) [inline src/make/ppx_make.ml]: In case you think the suggestion above would be helpful, here you could do the same thing:
```suggestion
      List.fold_left (fun return_ty (arg_label, arg_ty) ->
       ptyp_arrow ~arg_label lab arg_ty return_ty) body_ty arg_tys 
```

File: src/make/ppx_make.ml [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+    | Optional _ -> true
+    | _ -> false) labels
+    ;;
+  end
+
+   module Construct = struct
+     (* Additional AST construction helpers *)
+   
+     let apply_type ~loc ~ty_name ~tps = 
+      ptyp_constr ~loc (Located.lident ~loc ty_name) tps
```
Comment (pitag-ha) [inline src/make/ppx_make.ml]: Ooooh, I've just noticed that `tps` stands for **t**ype **p**arameter**s** and not for **t**y**p**e**s**, right? Might it make sense to spell that out in case someone else like me comes along and needs a bit clearer hints of what the variables stand for? And also, I think it would be more common to talk about the "return type" rather than the "apply type" (unless I'm misunderstanding what this function is doing). So what do you think about the following?
```suggestion
     let return_type ~loc ~ty_name ~ty_params = 
      ptyp_constr ~loc (Located.lident ~loc ty_name) ty_params
```

File: src/make/ppx_make.ml [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+           }
+         =
+         td
+       in
+       let tps = List.map (fun (tp, _variance) -> tp) ptype_params in
+       match ptype_kind with
+       | Ptype_record label_decls ->
+        if private_ = Public then  
+         let derived_item = create_make_sig ~loc ~ty_name ~tps label_decls in
+        derived_item
```
Comment (pitag-ha) [inline src/make/ppx_make.ml]:
```suggestion
        create_make_sig ~loc ~ty_name ~tps label_decls
```

File: src/make/ppx_make.ml [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+      else 
+        [psig_extension ~loc 
+          (Location.error_extensionf ~loc "We cannot expose functions that explicitly create private records.") [] ]
+       | _ -> []
```
Comment (pitag-ha) [inline src/make/ppx_make.ml:136]: If I understand correctly, this case is never supposed to be hit. Is that right? If so, I think instead of silently not doing anything, it would be better to return an error (by embedding the error as an extension node like always). It could be something along the lines "There has been an internal error while encountering a non-record type".

**Review by pitag-ha — COMMENTED** [review]
Thanks a lot for this implementation @ayc9 , it looks great! And sorry for the delay in reviewing (as you know, I was on vacation).

I only have one important question: why is this still marked as a draft? :) I think it's ready/very close to being ready to be merged! I've only looked over the first part so far and have left some minor comments; all of that is just nitpicking and nothing is important/blocking (in fact, you don't even need to take the suggestions if you don't agree).

Just one thing that I find a bit more important: It would be good to add something like "deriver make:" at the beginning of each error message. For example "deriver make: "We cannot expose functions that explicitly create private records.". That's just to make clear that the error doesn't come from the compiler, but from the deriver, when the user encounters it. Btw, it looks really good how you've embedded the errors!

P.S.: One thing that has called my attention: if I understand correctly, `find_main` does not only separate the arguments into the main argument and the other arguments but also turns around the order of the other arguments (which is needed later). Is that right? If so, it might be nice to somehow make that clear in the code to avoid confusion. I might come back to that later.

File: test/make/signature.t [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+  $ driver test.mli
+  type c = (int * int)[@@deriving make]
+  include
+    sig
+      [@@@ocaml.warning "-32"]
+      [%%ocaml.error
+        "Unsupported use of make (you can only use it on records)."]
+    end[@@ocaml.doc "@inline"][@@merlin.hide ]
+
+Test 4: Given a private type, embed error
```
Comment (pitag-ha) [inline test/make/signature.t]: I think we talked about this before and I just said "just fail on private types", but now I'm wondering: why would we fail on private types? Declaring a type private and using the make deriver on it seems like a good option to have a pretty much "abstract" type and deriving a make function for it. Let me know if what I'm saying is unclear!

File: test/make/signature.t [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    z: e }[@@deriving make]
+  include
+    sig
+      [@@@ocaml.warning "-32"]
+      val make_e : v:f -> w:bool -> e
+      val make_f : x:int -> y:bool -> z:e -> f
+    end[@@ocaml.doc "@inline"][@@merlin.hide ]
+
+Test 6: Given recursive types with at least one 
+record type, expose 1 make function for each type
```
Comment (pitag-ha) [inline test/make/signature.t]:
```suggestion
record type, expose 1 make function for each record 
```

Did you mean "record"? It's possible that I'm missing something here.

**Review by pitag-ha — APPROVED** [review]
Hey @ayc9 , once again: thanks for this great PR! I've just had a look at the tests as well and they all also look good, so I'm already approving the PR. There's one kind of test that might be worth adding: records with type parameters, i.e. something along the lines
```
type 'a t = {a: 'a option } [@@deriving make]
```
but like all the other comments, that's nothing important/blocking. Let me know if you want to look into some of the comments or if you want me to merge like this :)

File: src/make/ppx_make.ml (hunk 1 above)
Comment (ayc9) [reply src/make/ppx_make.ml]: No reason. I can undo that!

File: src/make/ppx_make.ml (hunk 2 above)
Comment (ayc9) [reply src/make/ppx_make.ml]: Yes that's definitely a better place for it! Will move it

File: src/make/ppx_make.ml (hunk 3 above)
Comment (ayc9) [reply src/make/ppx_make.ml]: Yes definitely -- good to know!

File: src/make/ppx_make.ml (hunk 4 above)
Comment (ayc9) [reply src/make/ppx_make.ml:38]: This line is from previous code, but it's interesting to note! Do you think about how likely it is that a release would change a certain type when making such a decision? How to recognize code "without a stability guarantee"?

File: src/make/ppx_make.ml (hunk 7 above)
Comment (ayc9) [reply src/make/ppx_make.ml]: Wow was also thinking of tps as types ... will rename!
I was understanding apply type as: apply (verb) the type derivation. `return_type` makes better sense bc it returns the derived type.

File: src/make/ppx_make.ml (hunk 9 above)
Comment (ayc9) [reply src/make/ppx_make.ml:136]: Checking for non-record types happens earlier in the is_derivable function, and there's a corresponding error: "Unsupported use of make (you can only use it on records)" @pitag-ha

File: test/make/signature.t (hunk 11 above)
Comment (ayc9) [reply test/make/signature.t]: Yes totally -- will fix!

File: test/make/signature.t (hunk 10 above)
Comment (ayc9) [reply test/make/signature.t]: What's failing here is the signature deriver -- the deriver will work with the structure of private types (see test 4 in `structure.t`. So the error is about exposing the private types. But are you saying it still shouldnt fail for the signature? @pitag-ha

File: src/make/ppx_make.ml (hunk 9 above)
Comment (ayc9) [reply src/make/ppx_make.ml:136]: ^ I'm tagging you to ask if we should still add that error msg if there is a similar one we explicitly check for earlier

Comment (pitag-ha) [reply src/make/ppx_make.ml:136]: It's up to you. I'd still add that error message for the following reason: if at some point in the future the code is changed and, in particular, the part of the code checking for non-record types earlier is removed (i.e. a small "bug" is introduced), then it would be good to have an error message that makes clear that there was an unexpected error in the deriver. With the current implementation, the deriver will simply ignore that type. Given that the deriver isn't supposed to be used on non-records in the first place though, it's also ok to leave it as is, if you prefer that.

File: test/make/signature.t (hunk 10 above)
Comment (pitag-ha) [reply test/make/signature.t]: Tbh, I'm not very familiar with private types myself. As far as I remember, it's possible and ok to expose private types. If that's the case, I think it would be nice to be able to apply the deriver to that. If you want, we can look into what exactly private types are together one of these days!

Comment (ayc9) [reply test/make/signature.t]: Ok! I just removed the private types check and promoted the tests. There is no longer a special case for private types

File: src/make/ppx_make.ml (hunk 9 above)
Comment (ayc9) [reply src/make/ppx_make.ml:136]: Ok just tried to change it and understood the need for `[]` in the tests with recursive types where there is a type that is not a record type: as long as at least 1 of the recursive types is a record, the deriver should still work on that record type. 

This point in the code is the case looking at the other non-record types in the recursive definition, and we don't actually want to throw an error.

Comment (ayc9) [reply src/make/ppx_make.ml:136]: Left it as is but maybe a clarifying comment might be useful?

Comment (pitag-ha) [reply src/make/ppx_make.ml:136]: > and understood the need for [] in the tests with recursive types where there is a type that is not a record type: as long as at least 1 of the recursive types is a record, the deriver should still work on that record type.

Oooh, that's a very good point! Nice, your super complete test base has already proven very useful!

> Left it as is but maybe a clarifying comment might be useful?

Yes, I think that might be nice. Thanks!

File: test/make/signature.t (hunk 10 above)
Comment (pitag-ha) [reply test/make/signature.t]: Cool, thanks!

File: src/make/ppx_make.ml (hunk 4 above)
Comment (pitag-ha) [reply src/make/ppx_make.ml:38]: > how likely it is that a release would change a certain type when making such a decision? How to recognize code "without a stability guarantee"?

The source code of OCaml is 100% stable. So it's safe to say that you always have a stability guarantee, unless you're using something that explicitly says it's unstable. Here, youre manipulating the AST, which comes from the compiler `Parsetree` module, which explicitly has a [warnings saying that it's unstable](https://ocaml.org/api/compilerlibref/Parsetree.html).
